### ISSUE TITLE
FITS Binary Table to Astropy Table round trip re-orders COMMENT cards

### ISSUE THREAD

**Issue by SaOgaz** [astropy/astropy#7530]
I'm really not sure if this could be considered a true bug or not, but I imagine it will cause some users some frustration.  When a FITS Binary file is round-tripped from an Astropy Table back to a FITS file (using `Astropy.table.Table.read` and `fits.table_to_hdu`) the `COMMENT` cards all get shuffled to the bottom of the header.  A lot of HST data uses these `COMMENT` cards like section titles, not sure if other data tends to do this.  I'm using astropy 3.1.dev21825 in this example.

Original, near the top of the header:
<img width="808" alt="screen shot 2018-06-04 at 12 06 35 pm" src="https://user-images.githubusercontent.com/5287848/40928679-336e56a2-67f0-11e8-9f18-613ab86bcefe.png">

Output after round tripping, at the bottom of the header:
<img width="835" alt="screen shot 2018-06-04 at 12 06 17 pm" src="https://user-images.githubusercontent.com/5287848/40928684-3aacc6f6-67f0-11e8-8da4-055e415ba9eb.png">

**Comment by saimn** [comment]
I agree it's a bug, the order of cards should be preserved (and there is probably the same issue with BLANK and HISTORY cards ?). Something to try would be to change `table_hdu.header.append((key, item))` to `table_hdu.header.append((key, item), end=True)` here:
https://github.com/astropy/astropy/blob/master/astropy/io/fits/convenience.py#L554

There is another issue here: the cards comment are lost. And since `Table.meta` is a simple dict-like object, it is not so easy to fix...

**Comment by pllim** [comment]
Wait a minute, isn't `dict` ordered by insertion in Python 3.6? Does this issue fixes itself in Python 3.6? @SaOgaz , which Python version are you using?

**Comment by saimn** [comment]
`Table.meta` is an `OrderedDict` by default I think. The issue has more to do with the way COMMENT/HISTORY/BLANK are managed by the `fits.Header` code (by default normal cards are inserted before these cards I think, hence the use of `end=True`).

**Comment by SaOgaz** [comment]
fyi, @saimn, so far I've gotten the feedback from my users that the keyword card comment loss is the lowest priority of the bugs I've found, so will probably ignore that for now.  

`table_hdu.header.append((key, item), end=True)` didn't work, but I think I'm not surprised, since the problem was them getting tacked onto the end in the first place.  Going to keep poking around in the code.

**Comment by SaOgaz** [comment]
Okay, actually, the more I look at this the more I'm stumped on how to handle it.  I was thinking we could use plain indexing numbers to keep track of everything, but it's not just the COMMENT and HISTORY cards that get re-ordered, it's also some of the regular keywords.  I think this is just the nature of how the header is getting re-populated on it's trip back to an HDU (for stuff not carried in the `meta` dict. But, this means that any index numbers that we could save out from the original HEADER/COMMENT keywords may no longer be a relevant ordering...

**Comment by SaOgaz** [comment]
After some discussion with @pllim, we realized that maybe keeping the original order of the header doesn't even make sense.  What do you do when there are new keywords that weren't there before? You could tack any new keywords onto the end of the header, but in the particular situation I was looking at, re-inserting COMMENT cards that are used as section titles might no longer be accurate section titles when new keywords (new columns etc) are being added.  You could have a section for `COMMENT        Column Formats` that doesn't have all the column formats because the new ones are at the bottom.  In that regard, it makes more sense to re-make the header the way it's being done now, with all keyword organized in a sensible way.  @saimn thoughts?

**Comment by saimn** [comment]
Actually the main issue is that COMMENT and HISTORY cards are put in a list (`tbl.meta['comments']`), thus the ordering of comments relatively to the other cards is lost.

> we realized that maybe keeping the original order of the header doesn't even make sense

I think it does make sense. You don't want your header to be shuffled by astropy when reading and writing a table.

>  What do you do when there are new keywords that weren't there before? You could tack any new keywords onto the end of the header...

`fits.Header` has several ways to handle this, typically a new card is added before the first comment, unless you use `end=True`:
https://gist.github.com/saimn/4b9a8c79c251874e35a7553d1ae900fb

So, obviously it is not possible to map a `Header` object to a simple dict. I wonder if we could just use the `Header` itself for the `.meta` attribute. Or maybe a "light" version of `fits.Header`. `.meta` just needs to be a *dict-like* object (http://docs.astropy.org/en/latest/table/construct_table.html#meta), so that could work.

**Comment by SaOgaz** [comment]
This seems like it would take a substantial update to `table_to_hdu` if I'm understanding your suggestion correctly? Because a lot of the column definition keywords (TTYPEn, TFORMn, etc) get mapped back out from `column` attributes.  So as those were getting put back into the new header, we'd have to re-order stuff according to the original Header (which seems like it might be a bit of an indexing nightmare).

This *was* one of the possible solutions @pllim and I discussed.  But the way the `COMMENT` cards are used in my test file would still not be happy with this solution, bringing up the point that the way things in any input header are completely arbitrarily organized.

It seems to come down to the question of what is more sensible, stick to the original header order as much as possible, knowing that it could result in any new column def keywords being put in some arbitrary position in the header (i.e., not necessarily grouped with the other existing column def keywords).  Or organize the column keywords in the new header in a "sensible way", and move any non-column keywords to the later portion of the header (which seems to be what's happening now).

**Comment by saimn** [comment]
> This seems like it would take a substantial update to table_to_hdu if I'm understanding your suggestion correctly?

It could be relatively easy if we just put the fits Header as the `.meta` attribute (or a simplified version to keep only the dict-like API). It could be optional also, with a new option.

> Because a lot of the column definition keywords (TTYPEn, TFORMn, etc) get mapped back out from column attributes. So as those were getting put back into the new header, we'd have to re-order stuff according to the original Header (which seems like it might be a bit of an indexing nightmare).

You never have to manage manually the structural keywords *(TTYPEn, TFORMn, etc)*, the fits Header/BinTableHDU already takes care of this, and keep them ordered properly (or at least it tries to). 
But if we try to use directly a fits Header as a Table meta, I think we should remove these keywords to avoid confusion. In this case the column attributes should be accessed/modified only through the Table's Column API, and this would be mapped back to fits attribute when creating an HDU.

**Comment by stscicrawford** [comment]
If the keywords are being kept in an order dict, one possibility would be to have an option that the table key words are written out at the same index as they would be read in at.   I'm not sure how easy this would be, but it might help at least help preserve some of the ordering.

**Comment by pllim** [comment]
Keeping the order (in a way that makes sense) gets complicated when header changes, e.g., add/remove keyword/comment/history/column. Sara and I talked about this and at some point concluded that it is easier to just show people how to re-order their header via some manual scripting. :grimacing:
